File: car.py
import time

class FakeDatabase:
   def __init__(self):
       self.data = {
           1: {"make": "Toyota", "model": "Corolla", "year": 2010},
           2: {"make": "Ford", "model": "Fusion", "year": 2012},
           3: {"make": "Honda", "model": "Civic", "year": 2015},
       }


   def get_car(self, car_id):
       # Simulate a slow database call
       time.sleep(2)
       return self.data.get(car_id, None)

class CarService:
   cache = {}


   def __init__(self):
       self.db = FakeDatabase()


   def get_car_info(self, car_id):
       if car_id in CarService.cache:
           print(f"Cache hit for car_id: {car_id}")
           return CarService.cache[car_id]
       else:
           print(f"Cache miss for car_id: {car_id}")
           car = self.db.get_car(car_id)
           if car is not None:
               CarService.cache[car_id] = car
               return car
           else:
               print(f"No car found in database for car_id: {car_id}")
               return "Car not found"


   def get_all_cars(self):
       for car_id in range(1, len(self.db.data) + 1):
           if car_id not in CarService.cache:
               self.get_car_info(car_id)
       return CarService.cache.values()


   def add_car(self, car_id, make, model, year):
       self.db.data[car_id] = {"make": make, "model": model, "year": year}
       print(f"Car added with id {car_id}: {make} {model} {year}")




# Example usage
service = CarService()
print(service.get_car_info(1))  
print(service.get_car_info(1))  
print(service.get_all_cars())  
service.add_car(4, "Nissan", "Altima", 2020) 
print(service.get_car_info(4))

File: car.js
class FakeDatabase {
    constructor() {
      this.data = {
        1: { make: "Toyota", model: "Corolla", year: 2010 },
        2: { make: "Ford", model: "Fusion", year: 2012 },
        3: { make: "Honda", model: "Civic", year: 2015 },
      };
    }
  
    // Simulate a slow database call
    getCar(carId) {
      return new Promise((resolve) => {
        setTimeout(() => {
          resolve(this.data[carId] || null);
        }, 2000); // 2 seconds delay
      });
    }
  }
  
  class CarService {
    static cache = {};

    constructor() {
      this.db = new FakeDatabase();
    }
  
    async getCarInfo(carId) {
      if (this.cache[carId]) {
        console.log(`Cache hit for car_id: ${carId}`);
        return this.cache[carId];
      } else {
        console.log(`Cache miss for car_id: ${carId}`);
        const car = await this.db.getCar(carId);
        if (car) {
          this.cache[carId] = car;
          return car;
        } else {
          console.log(`No car found in database for car_id: ${carId}`);
          return "Car not found";
        }
      }
    }
  
    async getAllCars() {
      const promises = [];
      for (let carId = 1; carId <= Object.keys(this.db.data).length; carId++) {
        if (!this.cache[carId]) {
          promises.push(this.getCarInfo(carId));
        }
      }
      await Promise.all(promises);
      return Object.values(this.cache);
    }
  
    addCar(carId, make, model, year) {
      this.db.data[carId] = { make, model, year };
      console.log(`Car added with id ${carId}: ${make} ${model} ${year}`);
    }
  }
  
  // Example usage
  const service = new CarService();
  
  (async () => {
    console.log(await service.getCarInfo(1)); // Cache miss
    console.log(await service.getCarInfo(1)); // Cache hit
    console.log(await service.getAllCars());  // Retrieve all cars
    service.addCar(4, "Nissan", "Altima", 2020); // Add a new car
    console.log(await service.getCarInfo(4)); // Cache miss, fetch from DB
  })();
  